server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/QueryToFilterAdapter.java

```diff
@@ -134,23 +127,26 @@ QueryToFilterAdapter<?> union(Query extraQuery) throws IOException {
          */
         extraQuery = searcher().rewrite(new ConstantScoreQuery(extraQuery));
         Query unwrappedExtraQuery = unwrap(extraQuery);
-        if (unwrappedExtraQuery instanceof MatchAllDocsQuery) {
-            return this;
-        }
         Query unwrappedQuery = unwrap(query);
         if (unwrappedQuery instanceof PointRangeQuery && unwrappedExtraQuery instanceof PointRangeQuery) {
             Query merged = MergedPointRangeQuery.merge((PointRangeQuery) unwrappedQuery, (PointRangeQuery) unwrappedExtraQuery);
             if (merged != null) {
                 // Should we rewrap here?
-                return new QueryToFilterAdapter<>(searcher(), key(), merged);
+                return new QueryToFilterAdapter(searcher(), key(), merged);
             }
         }
         BooleanQuery.Builder builder = new BooleanQuery.Builder();
         builder.add(query, BooleanClause.Occur.FILTER);
         builder.add(extraQuery, BooleanClause.Occur.FILTER);
-        return new QueryToFilterAdapter<>(searcher(), key(), builder.build()) {
+        Query rewrittenUnion = searcher().rewrite(new ConstantScoreQuery(builder.build()));
+        if (rewrittenUnion instanceof ConstantScoreQuery) {
+          rewrittenUnion = ((ConstantScoreQuery) rewrittenUnion).getQuery();
+        }
+        // This union is inefficient if Lucene cannot merge clauses of the boolean query through a rewrite.
+        final boolean inefficientUnion = rewrittenUnion instanceof BooleanQuery;
+        return new QueryToFilterAdapter(searcher(), key(), rewrittenUnion) {
             public boolean isInefficientUnion() {
-                return true;
+                return inefficientUnion;
```

Comment:
Ah. I see. You are letting the rewrite logic remove the match_all and match_none. That makes sense.